fastseq/optimizer/transformers/beam_search_optimizer.py

```diff
@@ -3,67 +3,138 @@
-from torch.nn import functional as F
-
-from transformers.generation_utils import (BeamHypotheses, GenerationMixin,
-                                           calc_banned_bad_words_ids,
-                                           calc_banned_ngram_tokens,
-                                           top_k_top_p_filtering)
+from torch import nn
+import torch.distributed as dist
+
+from transformers.file_utils import ModelOutput
+from transformers.generation_beam_search import (
+    BeamScorer, 
+    BeamSearchScorer
+)
+from transformers.generation_utils import (
+    GenerationMixin,
+    GreedySearchOutput,
+    SampleOutput,
+    BeamSearchOutput,
+    BeamSampleOutput,
+    StoppingCriteriaList,
+    BeamSearchEncoderDecoderOutput,
+    BeamSearchDecoderOnlyOutput
+)
+from transformers.models.bart.modeling_bart import BartForConditionalGeneration
+from transformers.models.prophetnet.modeling_prophetnet import ProphetNetForConditionalGeneration
+from transformers.models.t5.modeling_t5 import T5ForConditionalGeneration
+from transformers.models.gpt2.modeling_gpt2 import (
+    GPT2Model,
+    GPT2LMHeadModel,
+    GPT2DoubleHeadsModel
+)
+from transformers.generation_logits_process import (
+    _get_ngrams, 
+    _calc_banned_ngram_tokens, 
+    _get_generated_ngrams,
+    NoRepeatNGramLogitsProcessor,
+    EncoderNoRepeatNGramLogitsProcessor,
+    ForcedBOSTokenLogitsProcessor,
+    ForcedEOSTokenLogitsProcessor,
+    HammingDiversityLogitsProcessor,
+    InfNanRemoveLogitsProcessor,
+    LogitsProcessorList,
+    MinLengthLogitsProcessor,
+    NoBadWordsLogitsProcessor,
+    PrefixConstrainedLogitsProcessor,
+    RepetitionPenaltyLogitsProcessor,
+)
+from transformers.generation_stopping_criteria import (
+    StoppingCriteriaList,
+    validate_stopping_criteria,
+)
 
-from transformers.modeling_bart import BartForConditionalGeneration
-from transformers.modeling_t5 import T5ForConditionalGeneration
-from transformers.modeling_gpt2 import GPT2Model, GPT2LMHeadModel, GPT2DoubleHeadsModel
 from fastseq.ops.ngram_repeat_block import NGramRepeatBlock
-
 from fastseq.logging import get_logger
 from fastseq.utils.api_decorator import replace
 
 logger = get_logger(__name__, logging.INFO)
 
 no_repeat_ngram_op = NGramRepeatBlock()
 
-@replace(calc_banned_ngram_tokens)
-def calc_banned_ngram_tokens_v2(prev_input_ids: Tensor,
-                                num_hypos: int,
-                                no_repeat_ngram_size: int,
-                                cur_len: int,
-                                pad_token_id: int) -> None:
-    """Copied from fairseq for no_repeat_ngram in beam_search"""
-
-    if cur_len + 1 < no_repeat_ngram_size:
-        # return no banned tokens if we haven't generated no_repeat_ngram_size
-        # tokens yet
-        return [[] for _ in range(num_hypos)]
+@replace(_get_ngrams)
+def _get_ngrams_v2(ngram_size: int, prev_input_ids: torch.Tensor, num_hypos: int, pad_token_id: int = None):
```

Comment:
Is there any annotation for the return type?

Reply:
added one